### DISCUSSION TITLE
客户端和服务端双向认证的连接

### DISCUSSION THREAD

**Discussion by ValorBao** [Tencent/tquic#123] [Q&A (问题和回答)] [answered]
如何在客户端设置证书和tls版本，我使用quic_endpoint_new设置config，config里面配置了ssl_ctx,但是在quic_endpoint_connect里面崩溃在ssl_set_min_proto_version

**Comment by xiaofei0800** [comment, marked as answer]
[C示例参考](https://github.com/tquic-group/tquic-example-c)
服务端：针对SSL_CTX配置客户端认证并加载认证所需的根证书（在simple_server.c server_load_ssl_ctx函数中添加）
```
SSL_CTX_set_verify(server->ssl_ctx, SSL_VERIFY_FAIL_IF_NO_PEER_CERT | SSL_VERIFY_PEER, NULL); // 启用客户端认证，如果客户端不携带证书则认证失败
SSL_CTX_load_verify_locations(server->ssl_ctx, "ca.crt", NULL); // 加载客户端认证所使用的根证书
```
客户端：针对SSL_CTX配置携带客户端证书（在simple_client.c client_load_ssl_ctx函数中添加）
```
SSL_CTX_use_certificate_chain_file(client->ssl_ctx, "cert_client.crt");
SSL_CTX_use_PrivateKey_file(client->ssl_ctx, "cert_client.key", SSL_FILETYPE_PEM);
```
TLS版本不需要设置。
可参考[boringssl文档](https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html) SSL_CTX_set_verify相关部分。

另外，你的证书和TLS版本如何设置的？

**Reply by ValorBao** [reply]
我的证书是先把证书写在一片内存中，然后使用PEM_read_bio_X509读取，在使用SSL_CTX_use_certificate

**Reply by xiaofei0800** [reply]
可以参考上文提到的设置方法，对SSL_CTX进行设置。
